docs/DesignDocs/SGX2-EDMM.md

```diff
@@ -0,0 +1,105 @@
+# SGX2 EDMM in Open Enclave SDK
+
+Intel SGX2 on Icelake CPUs supports enclave dynamic memory management (EDMM),
+which many applications of the Open Enclave SDK (OE) will want to take advantage
+of. This document outlines the strategy for making these features available to
+applications.
+
+The key mechanism to support the dynamic addition of memory pages is the
+`ENCLS[EAUG]` instruction, which allocates and zeroes a new memory page via
+communication with the SGX driver. Before the enclave uses the new page, it
+acknowledges its addition by issuing an `ENCLU[EACCEPT]` instruction for the
+respective page address. Further, the `ENCLS[EMODPR]` and `ENCLU[EMODPE]`
+instructions restrict or extend page permissions, respectively. Page type
+changes (including page removal/trimming) are performed by the `ENCLS[EMODT]`
+instruction.
+
+## Heap allocation
+
+We propose to add a new OCALL to OE enclaves which requests new memory pages
+from the SGX driver (via `ioctl`), combined with an enclave function that, upon
+return from the OCALL, accepts the new pages. This does not require handling of
+signals or page faults and is easily integrated with memory managers like the
+default dlmalloc, by pre-mapping a configurable maximum number of pages and
+tracking the number of currently mapped pages. For performance reasons, we
+propose to add a switchless version of the OCALL whenever
+`OE_ENCLAVE_SETTING_CONTEXT_SWITCHLESS` is enabled:
+
+    trusted {
+      public oe_result_t oe_add_pages(size_t count);
+      public oe_result_t oe_add_pages_switchless(size_t count) transition_using_threads;
+    };
+
+The pluggable allocator API then uses `oe_add_pages`, before allocating
+additional pages, if necessary, before calling into `oe_allocator_*alloc`. The
+maximum number of heap pages (and perhaps other parameters like page
+permissions) are passed to `oe_allocator_init`.
+
+Applications that run a full OS kernel, for instance SGX-LKL, need to know about
+the size and layout of their address space and can therefore effectively use
+dynamic region allocation, which the SGX driver supports by tracking memory
+regions in the SGX driver. It automatically grows the heap by allocating a
+configurable number of pages upon request for a particular (unmapped) page.
+Dynamic regions are most conveniently declared as an initialization parameter to
+`oe_allocator_init` again.
+
+The permissions (R/W/X) of pages can be modified dynamically. Permissions of
+mapped pages are restricted by the `ENCLS[EMODPR]` instruction, which is a
+privileged instruction (ring 0) that the host executes. Similarly, permissions
+are extended by the `ENCLU[EMODPE]` instruction, which the enclave executes.
+OE's current pluggable allocators do not support permission changes. We propose
+to add a single, easy to use function to change (both restrict and extend) the
```

Comment:
internal or public?  If public, can we make it more TEE-agnostic?

Reply:
I'd love to! For now I kept this SGX-specific mainly because I don't have a good enough understanding of other TEE's memory management strategies, but I'm sure they also support at least r/w/x flags in some way. I'll do some reading.

Reply:
@wintersteiger https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/OpteeIntegration.md has some good information about OP-TEE.